### PR TITLE
fixed a memory leak in ReplayProgressObserver

### PR DESCRIPTION
**Pull request by LukasPaczos**
### Description

On each route leg change, regardless of a source (new route, reroute, new waypoint reached), we were appending all new events to the list of old events. For long running instrumentation tests with a significant amounts of route resets, the memory allocated for the list of events would grow indefinitely.

For a ~40 miles route reset every ~5 seconds the memory footprint could grow by ~30MB every ~20 minutes, or so.

### Changelog

```
<changelog>Fixed a small memory leak in the `ReplayProgressObserver` which accumulated events whenever current route leg changed.</changelog>
```

### CODE REVIEW

**Comment by Guardiola31337** [comment]
`ui-robo-tests` threw an `OutOfMemoryError`

```
12-15 07:39:34.139: D/MonitoringInstr(12594): Handling an uncaught exception thrown on the thread main.
12-15 07:39:34.139: D/MonitoringInstr(12594): java.lang.OutOfMemoryError: Failed to allocate a 7596 byte allocation with 5128 free bytes and 5KB until OOM
12-15 07:39:34.139: D/MonitoringInstr(12594): 	at java.util.ArrayList.toArray(ArrayList.java:496)
12-15 07:39:34.139: D/MonitoringInstr(12594): 	at java.util.ArrayList.addAll(ArrayList.java:188)
12-15 07:39:34.139: D/MonitoringInstr(12594): 	at ga.q.u()
12-15 07:39:34.139: D/MonitoringInstr(12594): 	at ga.j.u()
12-15 07:39:34.139: D/MonitoringInstr(12594): 	at ga.m.r()
12-15 07:39:34.139: D/MonitoringInstr(12594): 	at ga.j.r()
12-15 07:39:34.139: D/MonitoringInstr(12594): 	at j9.a.I()
12-15 07:39:34.139: D/MonitoringInstr(12594): 	at j9.a.K()
12-15 07:39:34.139: D/MonitoringInstr(12594): 	at com.mapbox.navigation.examples.core.ReplayHistoryActivity.u0()
12-15 07:39:34.139: D/MonitoringInstr(12594): 	at com.mapbox.navigation.examples.core.ReplayHistoryActivity.b0()
12-15 07:39:34.139: D/MonitoringInstr(12594): 	at n7.s2.a()
12-15 07:39:34.139: D/MonitoringInstr(12594): 	at w6.g.n0()
12-15 07:39:34.139: D/MonitoringInstr(12594): 	at w6.g.Y()
12-15 07:39:34.139: D/MonitoringInstr(12594): 	at w6.g$c$a.j()
12-15 07:39:34.139: D/MonitoringInstr(12594): 	at ka.a.i()
12-15 07:39:34.139: D/MonitoringInstr(12594): 	at za.x0.run()
12-15 07:39:34.139: D/MonitoringInstr(12594): 	at android.os.Handler.handleCallback(Handler.java:739)
12-15 07:39:34.139: D/MonitoringInstr(12594): 	at android.os.Handler.dispatchMessage(Handler.java:95)
12-15 07:39:34.139: D/MonitoringInstr(12594): 	at androidx.test.espresso.base.Interrogator.loopAndInterrogate(Interrogator.java:10)
12-15 07:39:34.139: D/MonitoringInstr(12594): 	at androidx.test.espresso.base.UiControllerImpl.loopUntil(UiControllerImpl.java:7)
12-15 07:39:34.139: D/MonitoringInstr(12594): 	at androidx.test.espresso.base.UiControllerImpl.loopUntil(UiControllerImpl.java:1)
12-15 07:39:34.139: D/MonitoringInstr(12594): 	at androidx.test.espresso.base.UiControllerImpl.loopMainThreadForAtLeast(UiControllerImpl.java:7)
12-15 07:39:34.139: D/MonitoringInstr(12594): 	at androidx.test.espresso.action.MotionEvents.sendDown(MotionEvents.java:9)
12-15 07:39:34.139: D/MonitoringInstr(12594): 	at androidx.test.espresso.action.Tap.sendSingleTap(Tap.java:4)
12-15 07:39:34.139: D/MonitoringInstr(12594): 	at androidx.test.espresso.action.Tap.-$$Nest$smsendSingleTap(Tap.java)
12-15 07:39:34.139: D/MonitoringInstr(12594): 	at androidx.test.espresso.action.Tap$1.sendTap(Tap.java:1)
12-15 07:39:34.139: D/MonitoringInstr(12594): 	at androidx.test.espresso.action.GeneralClickAction.perform(GeneralClickAction.java:4)
12-15 07:39:34.139: D/MonitoringInstr(12594): 	at androidx.test.espresso.ViewInteraction$SingleExecutionViewAction.perform(ViewInteraction.java:2)
12-15 07:39:34.139: D/MonitoringInstr(12594): 	at androidx.test.espresso.ViewInteraction.doPerform(ViewInteraction.java:21)
12-15 07:39:34.139: D/MonitoringInstr(12594): 	at androidx.test.espresso.ViewInteraction.-$$Nest$mdoPerform(ViewInteraction.java)
12-15 07:39:34.139: D/MonitoringInstr(12594): 	at androidx.test.espresso.ViewInteraction$1.call(ViewInteraction.java:2)
12-15 07:39:34.139: D/MonitoringInstr(12594): 	at androidx.test.espresso.ViewInteraction$1.call(ViewInteraction.java:1)
12-15 07:39:34.139: D/MonitoringInstr(12594): 	at java.util.concurrent.FutureTask.run(FutureTask.java:237)
12-15 07:39:34.139: D/MonitoringInstr(12594): 	at android.os.Handler.handleCallback(Handler.java:739)
12-15 07:39:34.139: D/MonitoringInstr(12594): 	at android.os.Handler.dispatchMessage(Handler.java:95)
12-15 07:39:34.139: D/MonitoringInstr(12594): 	at android.os.Looper.loop(Looper.java:135)
12-15 07:39:34.139: D/MonitoringInstr(12594): 	at android.app.ActivityThread.main(ActivityThread.java:5221)
12-15 07:39:34.139: D/MonitoringInstr(12594): 	at java.lang.reflect.Method.invoke(Native Method)
12-15 07:39:34.139: D/MonitoringInstr(12594): 	at java.lang.reflect.Method.invoke(Method.java:372)
12-15 07:39:34.139: D/MonitoringInstr(12594): 	at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:899)
12-15 07:39:34.139: D/MonitoringInstr(12594): 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:694)
12-15 07:39:34.139: W/AndroidJUnitRunner(12594): An unhandled exception was thrown by the app.
```

Could this be related to the changes brought by this PR? 🤔

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/replay/MapboxReplayer.kt [hunk 1]
```diff
@@ -23,6 +23,15 @@ class MapboxReplayer {
 
     private val replayEventsObservers: MutableSet<ReplayEventsObserver> = mutableSetOf()
 
+    /**
+     * Returns whether the replay is active an playing provided events.
```
Comment (Guardiola31337) [inline libnavigation-core/src/main/java/com/mapbox/navigation/core/replay/MapboxReplayer.kt:27]:
```suggestion
     * Returns whether the replay is active and playing provided events.
```

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/replay/route/ReplayProgressObserver.kt [hunk 2]
```diff
@@ -11,21 +11,22 @@ import com.mapbox.navigation.core.trip.session.RouteProgressObserver
...
 ) : RouteProgressObserver {
 
-    private val replayRouteMapper = ReplayRouteMapper()
     private var currentRouteLeg: RouteLeg? = null
 
     /**
      * @param options allow you to control the driver and car behavior.
      * @return [ReplayProgressObserver]
      */
+    @Deprecated("Provide a mapper through construct and change its options when needed.")
```
Comment (kmadsen) [inline libnavigation-core/src/main/java/com/mapbox/navigation/core/replay/route/ReplayProgressObserver.kt:29]: This is unrelated to a leak right?

**Comment by kmadsen** [comment]
@LukasPaczos this doesn't look like a leak of memory.. You're adding a way to clear old accumulated events, which will also break the `seekTo` feature

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/replay/route/ReplayProgressObserver.kt (hunk 2 above)
Comment (kmadsen) [reply libnavigation-core/src/main/java/com/mapbox/navigation/core/replay/route/ReplayProgressObserver.kt:29]: what's the reason for deprecating here? this seems to only remove the ability to change the replay speed, but the pr makes it appear as if it has something to do with memory leaks 🤔

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/replay/MapboxReplayer.kt (hunk 1 above)
Comment (kmadsen) [reply libnavigation-core/src/main/java/com/mapbox/navigation/core/replay/MapboxReplayer.kt:27]: The "and playing provided events" is not exactly accurate. `isPlaying()` will return true under this scenario

Notice there are no events to play
``` kotlin
val mapboxReplayer = MapboxReplayer()
mapboxReplayer.play()
assertTrue(mapboxReplayer.isPlaying()
```

```suggestion
     * Returns true after calling [play] and false when the replayer is stopped
```

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/replay/MapboxReplayer.kt [hunk 3]
```diff
@@ -23,6 +23,15 @@ class MapboxReplayer {
...
+     *
+     * @see play
+     * @see stop
+     * @see clearEvents
+     */
```
Comment (kmadsen) [inline libnavigation-core/src/main/java/com/mapbox/navigation/core/replay/MapboxReplayer.kt:32]:
```suggestion
     * @see clearEvents
     * @see finish
     */
```

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/replay/history/ReplayEventSimulator.kt [hunk 4]
```diff
@@ -61,6 +62,8 @@ internal class ReplayEventSimulator(
         jobControl.job.cancelChildren()
     }
 
+    fun isPlaying() = simulatorJob?.isActive == true
```
Comment (kmadsen) [inline libnavigation-core/src/main/java/com/mapbox/navigation/core/replay/history/ReplayEventSimulator.kt:65]: +1 do like the idea to add this

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/replay/route/ReplayProgressObserver.kt [hunk 5]
```diff
@@ -49,8 +50,13 @@ class ReplayProgressObserver(
         if (routeProgressRouteLeg != null) {
             val replayEvents = replayRouteMapper.mapRouteLegGeometry(routeProgressRouteLeg)
             if (replayEvents.isNotEmpty()) {
+                val wasPlaying = mapboxReplayer.isPlaying()
+                mapboxReplayer.clearEvents()
```
Comment (kmadsen) [inline libnavigation-core/src/main/java/com/mapbox/navigation/core/replay/route/ReplayProgressObserver.kt:54]: this will cause a lot of side effects. it also is not a solution for really long routeLegs

another way to reduce the memory footprint of the replayer would be to put the `ReplayEvents(mutableListOf())` list into a file

Comment (LukasPaczos) [reply libnavigation-core/src/main/java/com/mapbox/navigation/core/replay/route/ReplayProgressObserver.kt:54]: The ReplayProgressObserver already produces a lot of side effects and controls which samples are being played. If left to its own devices, it would accumulate the events indefinitely, even though most of them are detached from the current context (for example when the route changes to a new one).

My understanding is that if someone uses the ReplayProgressObserver, they are accepting the fact that this class will control the events that the replayer works with.

The problem I see is that even if ReplayProgressObserver does not clear the events automatically, we'd have to communicate to developers when they should clear the events manually in this case to avoid the memory growth in some of the scenarios, but there is no good hook to do that without breaking what ReplayProgressObserver was doing.

> another way to reduce the memory footprint of the replayer would be to put the ReplayEvents(mutableListOf()) list into a file

Could you elaborate? I definitely don't think the changes in the PR are a silver bullet and I'd love to find a better way to keep the number of events generate by the ReplayProgressObserver in check.

Maybe the ReplayProgressObserver should only clean up the events that it added instead of clearing all of them?

Comment (kmadsen) [reply libnavigation-core/src/main/java/com/mapbox/navigation/core/replay/route/ReplayProgressObserver.kt:54]: > Could you elaborate? I definitely don't think the changes in the PR are a silver bullet and I'd love to find a better way to keep the number of events generate by the ReplayProgressObserver in check.

Yeah instead of holding the event in memory they would be queried from a file with a cursor. Interface is all the same, but instead of bloating memory you're using a file. Obviously more complicated, and requires a new dependency to get access to the file system.
```
- val replayEvents = ReplayEvents(mutableListOf())
+ val replayEvents = ReplayEvents(filePath)
```

So I think it would is better to clear the events when the DirectionsRoute changes. This one also "resets" replay to the device location. We're using this implementation in a few places so it's worth adding IMO.
``` kotlin
class ReplayRoutesObserver(
    val mapboxReplayer: MapboxReplayer,
    val context: Context
) : RoutesObserver {
    override fun onRoutesChanged(result: RoutesUpdatedResult) {
        if (result.routes.isEmpty()) {
            mapboxReplayer.clearEvents()
            MapboxNavigationProvider.retrieve().resetTripSession()
            mapboxReplayer.pushRealLocation(context, 0.0)
            mapboxReplayer.play()
        }
    }
}
```

The next issue is that we have to subscribe 2 tools to get basic usage. [There is progress for a component that handles all these things](https://github.com/mapbox/mapbox-navigation-android/blob/05748f2e4a01db1c0fb41a8999a5ee7cd8e69eac/qa-test-app/src/main/java/com/mapbox/navigation/qa_test_app/lifecycle/DropInReplayViewModel.kt#L40-L41) in the qa-test-app

Comment (kmadsen) [reply libnavigation-core/src/main/java/com/mapbox/navigation/core/replay/route/ReplayProgressObserver.kt:54]: > The ReplayProgressObserver already produces a lot of side effects and controls which samples are being played. If left to its own devices, it would accumulate the events indefinitely, even though most of them are detached from the current context (for example when the route changes to a new one).

Also agree here. It's mostly a testing tool and your proposal to clear each `RouteLeg` is good. When we're dealing with seekTo issues, then we could solve those issues with another tool.

Sorry that I'm changing my mind, but yeah.. Clearing for each `RouteLeg` has my 💚

**Comment by dudeuter** [comment]
Just wondering if this is still an issue?

**Comment by LukasPaczos** [comment]
I would imagine so but this can only really be impactful if there are new routes replayed over and over again without restarting the replayer/app. Given how unlikely that is, we never prioritized this PR.
